crates/net/downloaders/src/bodies/concurrent.rs

```diff
@@ -98,37 +99,61 @@ where
     ///     1. The number of non-empty headers in the batch equals requested.
     ///     2. The total number of headers in the batch (both empty and non-empty)
     ///        is greater than or equal to the stream batch size.
-    ///     3. There are no more headers in the database.
+    ///     3. Downloader reached the end of the range
     ///
     /// NOTE: The batches returned have a variable length.
     fn query_headers(
         &self,
-        start: BlockNumber,
+        range: Range<BlockNumber>,
         max_non_empty: u64,
-    ) -> Result<Vec<SealedHeader>, db::Error> {
-        let tx = self.db.tx()?;
+    ) -> DownloadResult<Option<Vec<SealedHeader>>> {
+        if range.start >= self.download_range.end {
+            return Ok(None)
+        }
 
         // Acquire cursors over canonical and header tables
+        let tx = self.db.tx()?;
         let mut canonical_cursor = tx.cursor_read::<tables::CanonicalHeaders>()?;
         let mut header_cursor = tx.cursor_read::<tables::Headers>()?;
 
+        // Non empty headers count
         let mut non_empty_headers = 0;
+
+        // Collection of results
         let mut headers = Vec::<SealedHeader>::default();
-        let mut canonical_entry = canonical_cursor.seek_exact(start)?;
-        while let Some((number, hash)) = canonical_entry {
-            let (_, header) =
-                header_cursor.seek_exact((number, hash).into())?.expect("database corrupted");
+
+        let mut current_block_num = range.start;
+
+        // Collect headers while
+        //      1. Current block number is in range
+        //      2. The number of non empty headers is less than maximum
+        //      3. The total number of headers is less than the stream batch size
+        while current_block_num < range.end &&
+            non_empty_headers < max_non_empty &&
+            headers.len() < self.stream_batch_size
+        {
+            // Find the block hash
+            let (number, hash) = canonical_cursor
+                .seek_exact(current_block_num)?
+                .ok_or(DownloadError::MissingHeader { block_number: current_block_num })?;
+            // Find the block number
+            let (_, header) = header_cursor
+                .seek_exact((number, hash).into())?
+                .ok_or(DownloadError::MissingHeader { block_number: number })?;
+
+            // If the header is not empty, increment the counter
```

Comment:
Previously, the downloader would select headers from the database without enforcing any requirements for data validity (contiguous headers, within download range etc.). The logic is modified to ensure that the requested headers exist in the database, they are within a provided range and are contiguous